File: ts/2022/day10.ts
import { main, ocr } from 'lib/advent';
import { Point, PointSet, toAscii } from 'lib/coords';
import { Iter, iter } from 'lib/iter';
import { lines } from 'lib/util';

function parse(lines: string[]) {
  return lines.map((line) => {
    const [instr, arg] = line.split(' ');
    return { instr, arg: arg ? Number(arg) : null };
  });
}

type Instruction = { instr: string; arg: number };
type Cycle = { cycle: number; X: number };
type CRT = PointSet;

function registerValues(instructions: Instruction[]): Iter<Cycle> {
  return iter(instructions)
    .scan(
      (prev, { instr, arg }) => {
        const X = prev[prev.length - 1];
        return [X, ...(instr === 'addx' ? [X + arg] : [])];
      },
      [1]
    )
    .flat()
    .map((X, i) => ({ X, cycle: i + 1 }));
}

function signalStrengths(instructions: Instruction[]): number {
  return registerValues(instructions)
    .filter(({ cycle }) => (cycle % 40) - 20 === 0)
    .map(({ X, cycle }) => X * cycle)
    .sum();
}

function drawCrt(instructions: Instruction[]): CRT {
  const toPoint = (i: number): Point => ({ x: i % 40, y: Math.floor(i / 40) });

  return registerValues(instructions).reduce((crt, { cycle, X }) => {
    const pixel = toPoint(cycle - 1);
    const sprite = toPoint(X);
    if (Math.abs(sprite.x - pixel.x) <= 1) crt.add(pixel);
    return crt;
  }, new PointSet());
}

main(
  (s) => signalStrengths(parse(lines(s))),
  (s) => ocr(toAscii(drawCrt(parse(lines(s)))), '4x6')
);

File: py/2022/day10.py
import math
from itertools import islice
from typing import Iterator

from aoc import main
from aoc.util import chunks, ocr


def parse(s: str) -> Iterator[int]:
    X = 1
    for line in s.splitlines():
        match line.split(" "):
            case ["noop"]:
                yield X
            case ["addx", value]:
                yield X
                yield X
                X += int(value)


def signal_strengths(X: Iterator[int], start=19, step=40):
    return sum(
        x * (i * step + start + 1) for i, x in enumerate(islice(X, start, 240, step))
    )


def crt(X: Iterator[int]):
    display = [" "] * 240
    for cycle, x in enumerate(X):
        if abs(math.fmod(x, 40) - math.fmod(cycle, 40)) < 2:
            display[cycle] = "#"
    return ["".join(row) for row in chunks(display, 40)]


if __name__ == "__main__":
    main(
        lambda s: signal_strengths(parse(s)),
        lambda s: ocr(crt(parse(s)), "4x6"),
    )
